formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoder.kt

```diff
@@ -177,10 +182,207 @@ internal class IndefiniteLengthCborWriter(cbor: Cbor, output: ByteArrayOutput) :
     override fun incrementChildren() {/*NOOP*/
     }
 
+    override fun encodeTags(tags: ULongArray) = tags.forEach { getDestination().encodeTag(it) }
+
+}
+
+// optimized indefinite length encoder
+internal class StructuredCborWriter(cbor: Cbor) : CborWriter(cbor) {
+
+    /**
+     * Tags and values are "written", i.e. recorded/prepared for encoding separately. Hence, we need a helper that allows
+     * for setting tags and values independently, and then merging them into the final [CborElement] at the end.
+     */
+    internal sealed class CborContainer(tags: ULongArray, elements: MutableList<CborElement>) {
+        protected val elements = elements
+
+        var tags = tags
+            internal set
+
+
+        open fun add(element: CborElement) = elements.add(element)
+        class Map(tags: ULongArray, elements: MutableList<CborElement> = mutableListOf()) :
+            CborContainer(tags, elements)
+
+        class List(tags: ULongArray, elements: MutableList<CborElement> = mutableListOf()) :
+            CborContainer(tags, elements)
+
+        class Primitive(tags: ULongArray) : CborContainer(tags, elements = mutableListOf()) {
+            override fun add(element: CborElement): Boolean {
+                require(elements.isEmpty()) { "Implementation error. Please report a bug." }
+                return elements.add(element)
+            }
+        }
+
+        fun finalize() = when (this) {
+            is List -> CborList(content = elements, tags = tags)
+            is Map -> CborMap(
+                content = if (elements.isNotEmpty()) IntRange(0, elements.size / 2 - 1).associate {
+                    elements[it * 2] to elements[it * 2 + 1]
+                } else mapOf(),
+                tags = tags
+            )
+
+            is Primitive -> elements.first().also { it.tags += tags }
+
+        }
+    }
+
+    private operator fun CborContainer?.plusAssign(element: CborElement) {
+        this!!.add(element)
+    }
+
+
+    private val stack = ArrayDeque<CborContainer>()
+    private var currentElement: CborContainer? = null
+
+    // value tags are collects inside beginStructure, so we need to cache them here and write them in beginStructure or encodeXXX
+    // and then null them out, so there are no leftovers
+    private var nextValueTags: ULongArray = ulongArrayOf()
```

Comment:
It's better to be a function with a name that'll give away the behavior.

Also, I'm not sure that reallocating array every time we need to append a tag is the best way to collect tags. `ProtobufTaggedBase` shows a more efficient approach, although for CBOR `tagsStack`-analogue should be nullable and has `null` value by default (I assume that elements are usually untagged).